nexus/src/db/datastore.rs

```diff
@@ -4294,33 +4300,34 @@ impl DataStore {
             .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
     }
 
-    /// Look up a device authorization request by `user_code`.
-    // TODO-security: authz
-    pub async fn device_auth_get_request(
+    /// Delete a device authorization request.
+    pub async fn device_auth_request_hard_delete(
         &self,
         opctx: &OpContext,
-        user_code: String,
-    ) -> LookupResult<DeviceAuthRequest> {
+        authz_request: &authz::DeviceAuthRequest,
+    ) -> DeleteResult {
+        opctx.authorize(authz::Action::Delete, authz_request).await?;
+
         use db::schema::device_auth_request::dsl;
-        dsl::device_auth_request
-            .filter(dsl::user_code.eq(user_code))
-            .filter(dsl::time_expires.gt(Utc::now()))
-            .select(DeviceAuthRequest::as_select())
-            .get_result_async(self.pool_authorized(opctx).await?)
+        diesel::delete(dsl::device_auth_request)
+            .filter(dsl::user_code.eq(authz_request.id()))
+            .execute_async(self.pool_authorized(opctx).await?)
             .await
-            .map_err(|e| {
-                // TODO-correctness: better error (not found)
-                public_error_from_diesel_pool(e, ErrorHandler::Server)
-            })
+            .map(|rows_deleted| assert_eq!(rows_deleted, 1))
+            .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
     }
 
-    /// Grant a device authorization token.
-    // TODO-security: authz
-    pub async fn device_auth_grant(
+    /// Create a device access token record. The token may already
+    /// be expired if the device auth flow was not completed in time.
+    pub async fn device_access_token_create(
         &self,
         opctx: &OpContext,
+        authz_user: &authz::SiloUser,
         access_token: DeviceAccessToken,
     ) -> CreateResult<DeviceAccessToken> {
+        assert_eq!(authz_user.id(), access_token.silo_user_id);
+        opctx.authorize(authz::Action::CreateChild, authz_user).await?;
```

Comment:
This check feels a little weird, since `DeviceAccessToken` isn't declared by `lookup_resource!` as a child of `SiloUser`. Suggestions for a better check would be welcome.

Reply:
We could create another synthetic resource analogous to `ConsoleSessionList` (`DeviceAccessTokenList`) but I admit the boilerplate-to-value ratio on those is getting worse...

To be clear, would we expect this function to be invoked by the user themselves or by Nexus's external-authenticator?  I assumed the latter.

Reply:
No, this function is called as part of the confirmation in the browser, so it's made by an authenticated user.